WalletWasabi/BranchNBound/BranchAndBound.cs

```diff
@@ -0,0 +1,169 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using WalletWasabi.Helpers;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.BranchNBound
+{
+	public class BranchAndBound
+	{
+		private readonly Random _random = new();
+
+		public BranchAndBound(List<Money> availableCoins)
+		{
+			Count = availableCoins.Count;
+			SortedUTXOs = availableCoins.OrderByDescending(x => x.Satoshi).Select(c => c.Satoshi).ToArray();
+		}
+
+		private enum NextAction
+		{
+			AandB,
+			BandA,
+			A,
+			B,
+			Backtrack
+		}
+
+		private long[] SortedUTXOs { get; }
+		private int Count { get; }
+
+		public bool TryGetExactMatch(Money target, [NotNullWhen(true)] out List<Money> selectedCoins)
+		{
+			selectedCoins = new List<Money>();
+
+			if (SortedUTXOs.Sum() < target)
+			{
+				return false;
+			}
+
+			try
+			{
+				if (Search(target.Satoshi, out long[] solution))
+				{
+					selectedCoins = solution.Where(c => c > 0).Select(c => Money.Satoshis(c)).ToList();
+
+					if (CalcEffectiveValue(selectedCoins) == target)
+					{
+						return true;
+					}
+				}
+
+				return false;
+			}
+			catch (Exception ex)
+			{
+				Logger.LogError("Couldn't find the right pair. " + ex);
+				return false;
+			}
+		}
+
+		public Money CalcEffectiveValue(List<Money> list)
+		{
+			Guard.NotNull(nameof(list), list);
+
+			Money sum = Money.Satoshis(0);
+
+			foreach (var item in list)
+			{
+				sum += item.Satoshi;        // TODO: effectiveValue = utxo.value − feePerByte × bytesPerInput
+			}
+
+			return sum;
+		}
+
+		private bool Search(long target, [NotNullWhen(true)] out long[] solution)
+		{
+			// Current effective value.
+			long effValue = 0L;
+
+			// Current depth.
```

Comment:
Maybe: 
```suggestion
			// Current depth (think of the depth in the recursive algorithm sense).
```

Reply:
I'm not sure, that we need those comments at all.
`int depth = 0` and `long effValue = 0L;` is self-explanatory for me. We are giving them a starting value and that's all.

Reply:
Ok.